### PR TITLE
Fe 46 manage bookings UI

### LINKED ISSUE #46: FE: Manage Bookings UI
This view is for pet sitters who have requests coming in and have the ability to manage their bookings. 

Follow mockup. Refer to "Manage bookings" in the spec document

The mockup should include: 
-Calendar view
\- active upcoming bookings

### PR DESCRIPTION
**Pull request by jonahtabb**
resolves #46 

### What this PR does (required):
- Creates a UI for sitters to manage their bookings

### Screenshots / Videos (front-end only):
image

### Any information needed to test this feature (required):
This page can be viewed at http://localhost:3000/manage-bookings

### Any issues with the current functionality (optional):
- Style within the booking cards needs further adjustment, which I will do when I can pull in actual data from the backend.
- Need to add support for start and end time on booking cards.

### CODE REVIEW

File: client/src/pages/ManageBookings/ManageBookings.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    if (isBookedDay) {
+      dayStyle.style.backgroundColor = theme.palette.primary.main;
+      dayStyle.disabled = false;
+    }
+    if (isSelectedDay) {
+      dayStyle.style.border = `2px solid ${theme.palette.grey[400]}`;
+    }
+    const dayComponentStyled = cloneElement(dayComponent, dayStyle);
+    return <div>{dayComponentStyled}</div>;
```
Comment (ezeilo-su) [inline client/src/pages/ManageBookings/ManageBookings.tsx:47]: Please use MUI `Box` component for this `div`

**Review by ezeilo-su — COMMENTED** [review]
UI looks good! Just a minor comment.

File: client/src/helpers/ManageBookings/formatBookingDate.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export default function formatBookingDate(date: Date): string {
+  const day = date.getDate();
+  const month = new Intl.DateTimeFormat('en-US', { month: 'long' }).format(date);
+  const year = date.getFullYear();
+  return `${day} ${month} ${year}, 8-9`;
```
Comment (ugur-akin) [inline client/src/helpers/ManageBookings/formatBookingDate.ts:5]: Is there a reason we hardcoded `8-9` here? Ideally, we should avoid hardcoding things in FE/BE tickets in order not to indroduce overhead for integration. More on this idea below as well 👍🏼

File: client/src/helpers/ManageBookings/index.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import checkIfBooked from './checkIfBooked';
+import formatBookingDate from './formatBookingDate';
+import sortBookingsByStartDate from './sortBookingsByStartDate';
+
+export { checkIfBooked, formatBookingDate, sortBookingsByStartDate };
```
Comment (ugur-akin) [inline client/src/helpers/ManageBookings/index.ts:5]: It's nice to see consistent hierarchy here 👍🏼 However I think it's a bit over the top to separate each helper function into its own file. A single file can definitely qualify to contain more than a single function so I would recommend to group these together in a single file instead. Granted, this is a bit preferential but it will definitely reduce overhead in managing the project 🙌🏼

File: client/src/helpers/ManageBookings/sortBookingsByStartDate.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { BookingRequests } from '../../pages/ManageBookings/types';
+
+export default function sortBookingsByStartDate(bookingRequests: BookingRequests): BookingRequests {
+  const sorted = [...bookingRequests].sort((bookingA, bookingB) => {
+    return bookingB.start.getDate() - bookingA.start.getDate();
+  });
+  return sorted;
+}
```
Comment (ugur-akin) [inline client/src/helpers/ManageBookings/sortBookingsByStartDate.ts:8]: This is a nice helper function, however, I would have loved to see it designed to be a bit more flexible as a bonus (e.g. an `asc/desc` distinction, maybe via a parameter, at the very least). Nice touch on making it a *pure function* though with the spread operator😎 💯

File: client/src/pages/ManageBookings/ManageBookings.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import { DatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
+import { MaterialUiPickersDate } from '@material-ui/pickers/typings/date';
+import { cloneElement, useState } from 'react';
+import { CurrentBookings, NextBooking, PastBookings } from '../../components/ManageBookings/';
+import { checkIfBooked } from '../../helpers/ManageBookings';
+import mockBookingData from './mockBookingData';
+import { BookingRequests } from './types';
+import useStyles from './useStyles';
+
+export default function ManageBookings(): JSX.Element {
```
Comment (ugur-akin) [inline client/src/pages/ManageBookings/ManageBookings.tsx:12]: This component is the perfect location to move over processing logic of Bookings other than a `React.Context`!

File: client/src/pages/ManageBookings/ManageBookings.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            <Paper elevation={2}>
+              <Box p={2} pt={2}>
+                <Typography variant="h6" component="div" className={classes.uppercase}>
+                  Current Bookings
+                </Typography>
+                <CurrentBookings bookingRequests={sitterBookings} />
+                <Typography variant="h6" component="div" className={classes.uppercase}>
+                  Past Bookings
+                </Typography>
+                <PastBookings bookingRequests={sitterBookings} />
```
Comment (ugur-akin) [inline client/src/pages/ManageBookings/ManageBookings.tsx:74]: We could have been more [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) here, both of these sections are almost identical. If you take a closer look, `PastBookings` and `CurrentBookings` components are exactly the same in representation (e.g. they return the same JSX) and when we end up moving the logic for processing bookings, we will end up with duplicate components. Instead we could componentize these sections by taking a list of bookings, a section title and a fallback (e.g. you have this in `NextBooking` ([`lines 38-40`](https://github.com/hatchways/team-gingerbread/blob/232ca0d94fc825c32cce29d44428ff9e0e560b21/client/src/components/ManageBookings/NextBooking.tsx#L38-L40)), I think it would be nice to have them for `Past/CurrentBookings` as well).

File: client/src/pages/ManageBookings/mockBookingData.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+const mockBookingData = [
+  {
+    description: 'Test Booking Request Description #1',
+    status: BookingStatus.pending,
+    requestId: '61a26c5dc517da0ca7a2b9',
+    userId: '619c203917b5e66160b119fb',
+    sitterId: '618ff3d3939f8555dc391646',
+    start: new Date(2021, 10, 9),
+    end: new Date(2021, 10, 11),
+  },
```
Comment (ugur-akin) [inline client/src/pages/ManageBookings/mockBookingData.ts:12]: First of all, well done using mocks 🎉 - it's a best practice that's overlooked often! This is a very good start, but we can improve on it: the format of a mock data shouldn't necessarily be identical to the backend model of what we are showing (e.g. `Request` in this case). Instead, it should be tailored to what the frontend expects - and it's important to realize this can vary from how we store things in the backend.

Let's consider our use case: 
- We need the details of a booking to filter and display it properly, so it makes sense to expect fields such as `status`, `description`, `start` and `end`.
- The mocks show us the booking cards need to display the user's name and profile picture, hence we should expect those. To me it makes more sense to expect a field such as `ownerProfile` instead of user id. I can't really work with the user id - if I can't display a user's name or profile picture with it, what good is it to me? Right?
- Since this page is for sitters only, sitter id is going to be my (logged in user's) user id (or profile id depending on BE implementation). I already have access to both, do I need to expose it to the frontend? Are we going to do something with it? (no I'm asking seriously, I might be missing something :D) It doesn't seem like my mock data needs to include that at a first glance. Besides I'm always against exposing user ids unnecessarily, goes for both user and sitter ids - you can never be too safe 😉

Clearly and explicitly conveying exactly what you expect in your mock data makes integration soo much simpler 😮If I were to integrate this page and just retrieved the bookings from the API like the mocks, I would soon realize I need to do more work on the FE (such as retrieve the owner's user and it's profile, then integrate them in the appropriate parts which involve more FE work). If your mocks are formatted exactly as you expect the data to be passed from the "API layer", then my only work would be to write the appropriate fetching logic (i.e either populate on the backend (ideal) or do a few fetches to populate the necessary fields manually). 

In summary, it should be frontend's job to identify what it needs from the BE and it should be integration's job to figure out how to get it!

Hopefully I wasn't too confusing here😅 Feel free to ask for further clarification if needed 👍🏼

File: client/src/pages/ManageBookings/types.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+export enum BookingStatus {
+  pending,
+  accepted,
+  declined,
+}
```
Comment (ugur-akin) [inline client/src/pages/ManageBookings/types.ts:5]: Very nice 🎉

**Review by ugur-akin — CHANGES_REQUESTED** [review]
Well done with this feature! I really liked the use of React/Material UI 👍🏼 I went ahead and left a few comments about issues that I would consider *more advanced* than average so don't let them overwhelm you - we can always leave some of the code as is and tackle the ideas introduced in the comments later 🙌🏼

Keep it up! :)

File: client/src/pages/ManageBookings/ManageBookings.tsx (hunk 1 above)
Comment (jonahtabb) [reply client/src/pages/ManageBookings/ManageBookings.tsx]: resolved by decf3104dfffa92f05635b7a287a416bcf074dd3
Actually changed the div to a react fragment, because div / box was unnessary.

File: client/src/components/ManageBookings/CurrentBookings.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const CurrentBookings: FC<CurrentBookingsProps> = ({ bookingRequests: bookingRequests }) => {
+  const currentBookings = bookingRequests.filter((bookingRequest) => {
+    const startDayEpoch = bookingRequest.start.getTime();
+    const today = new Date();
+    today.setHours(0, 0, 0, 0);
+    const todayUTC = today.getTime();
+    return startDayEpoch >= todayUTC;
+  });
```
[the remark this answers is not in the document]
Comment (jonahtabb) [reply client/src/components/ManageBookings/CurrentBookings.tsx]: Thank you super helpful.  resolved by 518715c87a7328daa88f1d060cb14634402290d3

File: client/src/components/ManageBookings/CurrentBookings.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+  const classes = useStyles();
+  const sortedBookings = sortBookingsByStartDate(currentBookings);
+  sortedBookings.pop();
```
[the remark this answers is not in the document]
Comment (jonahtabb) [reply client/src/components/ManageBookings/CurrentBookings.tsx]: I moved this to a high level, and refactored to use helper functions, thank you! resolved by ee1bbec1febc1f475166e3a5bf984d683305faf5

File: client/src/components/ManageBookings/NextBooking.tsx [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  const currentBookings = bookingRequests.filter((bookingRequest) => {
+    const startDayEpoch = bookingRequest.start.getTime();
+    const today = new Date();
+    today.setHours(0, 0, 0, 0);
+    const todayUTC = today.getTime();
+    return startDayEpoch >= todayUTC;
+  });
+
+  const sortedBookings = sortBookingsByStartDate(currentBookings);
+  const nextBooking = sortedBookings[sortedBookings.length - 1];
```
[the remark this answers is not in the document]
Comment (jonahtabb) [reply client/src/components/ManageBookings/NextBooking.tsx]: Thank you, makes sense.

[the remark this answers is not in the document]
Comment (jonahtabb) [reply client/src/components/ManageBookings/NextBooking.tsx]: resolved by ee1bbec1febc1f475166e3a5bf984d683305faf5

File: client/src/components/ManageBookings/PastBookings.tsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        <Box key={booking.requestId} className={classes.bookingContainer}>
+          <Typography variant="body2">{formatBookingDate(booking.start)}</Typography>
+          <Box display="flex" justifyContent="space-between">
+            <Box>
+              <Typography variant="body2">NAME OF DOG OWNER</Typography>
+            </Box>
+            <Box>{BookingStatus[booking.status]}</Box>
+          </Box>
+        </Box>
+      ))}
```
[the remark this answers is not in the document]
Comment (jonahtabb) [reply client/src/components/ManageBookings/PastBookings.tsx:37]: Got it thank you. ee1bbec1febc1f475166e3a5bf984d683305faf5

File: client/src/helpers/ManageBookings/checkIfBooked.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { BookingRequests } from '../../pages/ManageBookings/types';
+
+export default function checkIfBooked(bookingRequests: BookingRequests, dateToCheck: Date | null): boolean {
+  const check = bookingRequests.findIndex((booking) => {
```
[the remark this answers is not in the document]
Comment (jonahtabb) [reply client/src/helpers/ManageBookings/checkIfBooked.ts]: switched to isThereBookingOnThisDate, as I do find it helpful when functions that return a boolean begin with 'is'. resolved by 3eb549d2d991b584801ba1f3186d9ea83bbc74b1

File: client/src/helpers/ManageBookings/formatBookingDate.ts (hunk 2 above)
Comment (jonahtabb) [reply client/src/helpers/ManageBookings/formatBookingDate.ts:5]: No, this is a placeholder, and bad practice I imagine - I wanted to get the feature pushed.  I forgot that I would be needing to display time, so I hard-coded a placeholder.

File: client/src/pages/ManageBookings/ManageBookings.tsx (hunk 6 above)
Comment (jonahtabb) [reply client/src/pages/ManageBookings/ManageBookings.tsx]: I see, that makes a lot of sense.  The processing can happen in the high level component, and then all 3 display components can be instances of the same lower-level display component.

File: client/src/pages/ManageBookings/mockBookingData.ts (hunk 7 above)
Comment (jonahtabb) [reply client/src/pages/ManageBookings/mockBookingData.ts:12]: @ugur-akin This (and your other comments) have been super helpful.  Thank you for taking the time for this level of explanation. I am time-limited at the moment and am not going to fully revise the mock, but certainly will utilize this feedback in the future.

**Review by rajivtitus — APPROVED** [review]
Great work making the requested changes! Just a couple of comments are pending, please address them prior to merging this 👍🏽

File: client/src/components/ManageBookings/useStyles.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import { makeStyles } from '@material-ui/core/styles';
+
+const useStyles = makeStyles((theme) => ({
+  bookingContainer: {
+    minHeight: '75px',
+    border: `1px solid ${theme.palette.grey[200]}`,
+    padding: '10px',
+    marginBottom: '5px',
+  },
```
[the remark this answers is not in the document]
Comment (ababber) [reply client/src/components/ManageBookings/useStyles.ts:9]: I moved minHeight, padding, and marginBottom to the Box component in CurrentBookings, NextBookings, and PastBookings.

File: client/src/helpers/ManageBookings/index.ts (hunk 3 above)
Comment (ababber) [reply client/src/helpers/ManageBookings/index.ts]: I refactored all these separate files into named exports in /helpers/ManageBookings.tsx.
